internal/rpc/httpproxy.go

```diff
@@ -7,98 +7,117 @@ import (
+		ok := cp.AppendCertsFromPEM([]byte(ctl.Controller.CACertificate))
 		if !ok {
 			zapctx.Warn(ctx, "no CA certificates added")
 		}
 		tlsConfig = &tls.Config{
 			RootCAs:    cp,
-			ServerName: ctl.TLSHostname,
+			ServerName: ctl.Controller.TLSHostname,
 			MinVersion: tls.VersionTLS12,
 		}
 	}
 
-	if ctl.PublicAddress != "" {
-		err := doRequest(ctx, w, req, httpOptions{
-			TLSConfig: tlsConfig,
-			URL:       createURLWithNewHost(*req.URL, ctl.PublicAddress),
-		})
-		if err == nil {
-			return nil
-		}
+	// transport is the default HTTP transport config with TLS configuration.
+	transport := &http.Transport{
+		TLSClientConfig:       tlsConfig,
+		Proxy:                 http.ProxyFromEnvironment,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
 	}
-	for _, hps := range ctl.Addresses {
-		for _, hp := range hps {
-			err := doRequest(ctx, w, req, httpOptions{
-				TLSConfig: tlsConfig,
-				URL:       createURLWithNewHost(*req.URL, fmt.Sprintf("%s:%d", hp.Value, hp.Port)),
-			})
-			if err == nil {
-				return nil
-			} else {
-				zapctx.Error(ctx, "failed to proxy request: continue to next addr", zaputil.Error(err))
-			}
-		}
+
+	if len(urls) == 0 {
+		zapctx.Error(ctx, "no controller addresses found", zap.String("controller", ctl.Controller.Name))
+		http.Error(w, "no controller addresses found", http.StatusInternalServerError)
+		return
 	}
 
-	return errgo.New("couldn't reach a valid address for controller")
-}
+	if len(urls) > 1 {
+		rand.Shuffle(len(urls), func(i, j int) {
+			urls[i], urls[j] = urls[j], urls[i]
+		})
+	}
 
-func doRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, opt httpOptions) error {
-	client := &http.Client{
-		Transport: &http.Transport{
-			TLSClientConfig: opt.TLSConfig,
+	// TODO: Consider implementing a better load balancing mechanism that handles
+	// multiples URLs and handles failing backends gracefully e.g. try send to first
+	// URL and on failure, try second, etc.
+	proxy := &httputil.ReverseProxy{
+		Rewrite: func(pr *httputil.ProxyRequest) {
+			pr.SetURL(urls[0])
+			pr.Out.SetBasicAuth(names.NewUserTag(ctl.Username).String(), ctl.Password)
 		},
+		Transport: transport,
 	}
-	req = req.Clone(ctx)
-	req.RequestURI = ""
-	req.URL = &opt.URL
-	resp, err := client.Do(req)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
+	proxy.ServeHTTP(w, req)
+}
 
-	// copy headers
-	for k, vv := range resp.Header {
-		for _, v := range vv {
-			w.Header().Add(k, v)
+func getControllerAddresses(ctl dbmodel.Controller) ([]*url.URL, error) {
+	urls := make([]*url.URL, 0, 1)
+	if ctl.PublicAddress != "" {
+		address := ctl.PublicAddress
+		if !strings.Contains(address, "://") {
+			address = defaultScheme + "://" + address // ensure the address has a scheme
 		}
+		newURL, err := url.Parse(address)
```

Comment:
Shouldn't we be parsing the public address, and then doing:
```go
	u.Scheme = defaultScheme
```
What you're doing here is what url.Parse should do for us...

Reply:
So then we can just do u.Scheme = defaultScheme always?